feat(iam): `Role.withoutPolicyUpdates()`

rix0rrr opened the pull request:
Add a new method to `Role` called `withoutPolicyUpdates()`, which returns a wrapper object that will drop all policy updates. This can be used in situations where you want to skip the default IAM permission adding behavior of CDK.

Needs only be used with roles that are created in the same CDK application; for imported roles, supplying `mutable=false` when calling `Role.fromRoleArn()` is sufficient.

Fixes #2985.
Fixes #4465.
Closes #4501.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

<!-- 
Please read the contribution guidelines and follow the pull-request checklist:
https://github.com/aws/aws-cdk/blob/master/CONTRIBUTING.md
 -->

## Commit Message
feat(iam): `Role.withoutPolicyUpdates()`

Add a new method to `Role` called `withoutPolicyUpdates()`, which returns a wrapper object that will drop all policy updates. This can be used in situations where you want to skip the default IAM permission adding behavior of CDK.

Needs only be used with roles that are created in the same CDK application; for imported roles, supplying `mutable=false` when calling `Role.fromRoleArn()` is sufficient.

Fixes #2985.
Fixes #4465.
Closes #4501.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* Result: FAILED
* [Build Logs](https://dfq49x5zyi.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=066485e5-648d-4841-a24a-a11a6f550e5a%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* Result: FAILED
* [Build Logs](https://dfq49x5zyi.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=077ab841-9a20-4ab6-ba31-b1ac917fe054%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* Result: SUCCEEDED
* [Build Logs](https://dfq49x5zyi.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=65436077-980b-42b4-bab3-eca18b3ee4f3%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

skinny85 commented on packages/@aws-cdk/aws-iam/lib/policy.ts:
```diff
@@ -56,6 +56,22 @@ export interface PolicyProps {
+   *
+   * @default false
+   */
+  readonly mustCreate?: boolean;
```
I'm not crazy about this name... How about `lazy?: boolean`?

skinny85 commented on packages/@aws-cdk/aws-iam/lib/policy.ts:
```diff
@@ -160,19 +173,60 @@ export class Policy extends Resource implements IPolicy {
+        result.push('Policy created with mustCreate=true is empty. You must add statements to the policy');
+      }
+      if (!this.mustCreate && this.referenceTaken) {
+        result.push('Policy name has been read of empty policy. You must add statements to the policy so it can exist.');
```
I think we can word this a little more clearly... how about: `This Policy has been referenced by another resource, and so it must contain at least one statement`.

skinny85 commented on packages/@aws-cdk/aws-iam/lib/policy.ts:
```diff
@@ -160,19 +173,60 @@ export class Policy extends Resource implements IPolicy {
+        result.push(`Policy created with mustCreate=true must be attached to at least one principal: user, group or role`);
+      }
+      if (!this.mustCreate && this.referenceTaken) {
+        result.push('Policy name has been read of unattached policy. Attach to at least one principal: user, group or role.');
```
Same here: `This Policy has been referenced by another resource, and so it must be attached to at least one principal: user, group or role`.

skinny85 commented:
> I am not sure I fully understand the use case for this feature, and I couldn't dig those up from the referenced issues. #2985 is about immutable imported roles, #4465 is about maximum policy size limitations and I am not sure I fully understand how immutable roles help here.

The maximum policy size is related to the immutable roles because immutability decreases the size of the statements added to the role when it's passed into constructs that add statements to it. An example from #4465:

```ts
           const myRole = new iam.Role(this, 'Role', { ... });

           // use myRole in a bunch of places...

           // then:
           new codepipeline_actions.CodeBuildAction({
                // ...
                role: myRole,
           });
```

Now, it might happen that passing `myRole` to `CodeBuildAction` will add so many statements that it will exceed the limit. In that case, the customer could do:

```ts
           new codepipeline_actions.CodeBuildAction({
                // ...
                role: new iam.ImmutableRole(myRole),
           });
```

And this way, all statement-adding expressions in `CodeBuildAction` would be effectively ignored, and the customer could work around the "maximum policy size exceeded" error they get.

Hope the motivation for this is a little more clear now!

eladb commented:
README missing is "changes requested" and also I wanted to make sure the motivation for this, and the use case is clear before it's merged. 

To be honest, the motivation @skinny85 provided above doesn't sounds like a "proper" use case. Maybe it's some kind of an escape hatch to allow people to work around bugs or issues we have in the system, but then I'd argue that a whole new type of `Role` is not the right API.

Let's discuss this before we introduce additional public APIs that don't have a clear use case.

I think trying to write a README entry for this will be a good exercise as it will require you to articulate a reasonable situation where someone would want to use this feature.

rix0rrr commented:
I think we can chunk this PR smaller. There are some useful changes in there which are separate from the `ImmutableRole` class.

rix0rrr commented:
Let me try this again as https://github.com/aws/aws-cdk/pull/5608. Although that doesn't fix #4465.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* Result: SUCCEEDED
* [Build Logs](https://dfq49x5zyi.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=5c76776b-33e7-4b22-b3f0-95614d5a21bd%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

rix0rrr commented:
Added a README explaining why you might want to use this class.

Pre-emptive FAQ: 

## Can we not do anything else if the policy size is growing too big?

We need to apply judicious use of wildcards then, and we decide not to do IAM policy manipulation because it's scary. Therefore, we have to punt to the user in *some* way. This is that way.

We could make it slightly easier for the user to get an ImmutableRole, maybe instead of 

```ts
new codepipeline.Pipeline(this, 'Pipeline', {
  role: new iam.ImmutableRole(role),
});
```

they could do:

```ts
new codepipeline.Pipeline(this, 'Pipeline', {
  role: role.immutableCopy(),
});
```

But that's about the extent of it, and that's not appreciably different imo.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* Result: FAILED
* [Build Logs](https://dfq49x5zyi.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=eff18253-a413-4925-8af9-1b0238f80475%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* Result: FAILED
* [Build Logs](https://dfq49x5zyi.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=d650111d-522a-44c6-8bdc-8a288f65b856%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* Result: SUCCEEDED
* [Build Logs](https://dfq49x5zyi.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=27699db5-a877-4c43-acca-5cc8d5c1e6d3%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

eladb commented:
Thanks for adding the README and the FAQ, they are very helpful.

I think this API is much better:

```ts
new codepipeline.Pipeline(this, 'Pipeline', {
  role: role.immutableCopy(),
});
```

As it doesn’t introduce a whole new type and is more discoverable. You just return an IRole this way!. It can also be applied both to owned and imported roles, so there won’t be a difference (maybe even deprecate “mutable: false”).

I am still not sure about the terminology of immutable. Perhaps something more around “role.withoutPolicyUpdates()” or something around “disable automatic policy” which is more in line with the README narrative.

rix0rrr commented:
Mmyeah. I agree that a `role.withoutPolicyUpdates()` looks nicer at the usage site.

I don't buy adding it to the `IRole`:

* For one, it makes the interface unnecessarily heavy to implement. In Java-like languages (which we are roughly emulating using jsii) there is no way to say "classes that implement `IRole` *automatically* also have these operations available". We're left taking them as an argument, either showing a visible wrapper class (`new Immutable(role: IRole)`, or hiding the name of the wrapper class behind a factory function like `public static makeImmutable(role: IRole): IRole`, which is effectively the same thing).
* For two, I believe that for imports our default of `mutable: true` is probably wrong, and should be `mutable: false`. If you're using existing roles, you're probably looking to use them as-is (99% certainly because they're provisioned to you by your Ops team and you don't have permissions to create new roles), otherwise why wouldn't you be creating new roles? So if we were to put this `withoutPolicyUpdates()` on the `IRole` and rely on people calling that for imports, almost all people doing imports would be writing this:

```ts
const role = Role.fromRoleArn(....).withoutPolicyUpdates();
```

Which is a little silly.

My proposal is:

* Hide the `ImmutableRole` type and put it behind a method `withoutPolicyUpdates()` on the `Role` class.
* In a different PR, switch the `mutable` default of `fromRoleArn()` to `false`.

eladb commented:
@rix0rrr wrote:

> Hide the ImmutableRole type and put it behind a method withoutPolicyUpdates() on the Role class.

Sounds good.

> In a different PR, switch the mutable default of fromRoleArn() to false.

Yeah I agree that this makes more sense to be honest. And then I would deprecate the name `immutable` and rename this to use similar terminology (something like `policyUpdates: true|false`).

skinny85 commented:
> For two, I believe that for imports our default of mutable: true is probably wrong, and should be mutable: false. If you're using existing roles, you're probably looking to use them as-is (99% certainly because they're provisioned to you by your Ops team and you don't have permissions to create new roles), otherwise why wouldn't you be creating new roles? So if we were to put this withoutPolicyUpdates() on the IRole and rely on people calling that for imports, almost all people doing imports would be writing this:

Are you sure? This  (`mutable: false` - it wasn't called that way, but that's what it effectively was) was the behavior we had originally, and it was changed (to be `mutable: true` by default). I'm pretty sure we had a reason to make that change. Also, I'm not sure reverting to the previous behavior now is backwards-compatible...

skinny85 commented:
I personally think `withoutPolicyUpdates()` is just a roundabout way of saying "I want the role to be immutable", so I'd rather be direct than roundabout, but I don't feel super strongly about this.

rix0rrr commented:
> Are you sure? This (`mutable: false` - it wasn't called that way, but that's what it effectively was) was the behavior we had originally, and it was changed (to be `mutable: true` by default)

This is where we introduced the `mutable` keyword: https://github.com/aws/aws-cdk/pull/3716 and we went from a mutable imported Role to introducing the `mutable` keyword with default `true`.

We added support for adding policies to imported roles here: https://github.com/aws/aws-cdk/pull/2805. This is where I think we went wrong though.

Motivating issues for that:
https://github.com/aws/aws-cdk/issues/2651 - CodeBuild doesn't work with imported role: solved by not rendering unattached policy.
https://github.com/aws/aws-cdk/issues/2381 - Unable to use SQS event Source with imported role: actual root cause was using the wrong IAM action (but that only became visible when using imported roles)
(2 other issues are referenced that don't actually have anything to do with imported roles)

So it seems like the correct choice to me.

You're right the change is not backwards compatible but I think we'll do everyone a favor in the long run if we rip off the band-aid now. The flag is `@experimental` anyway.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* Result: FAILED
* [Build Logs](https://dfq49x5zyi.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=bd6f4878-fb9c-47ab-ba99-d5f6ae79d5f9%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* Result: SUCCEEDED
* [Build Logs](https://dfq49x5zyi.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=305ac43e-990b-4375-bf11-83779f3a06c8%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

skinny85 commented:
> #2651 - CodeBuild doesn't work with imported role: solved by not rendering unattached policy.

I actually don't agree that this was the error in the general case, as #2651 was a way to workaround an actual bug in CodeBuild (missing `DependsOn` between Policy and Project resources). If you switch this over now, everyone who has a working setup with CodeBuild in a VPC with an imported role will most likely be broken (as we will remove the required permissions CodeBuild needs to [connect to the VPC](https://github.com/aws/aws-cdk/blob/01220cf2139ea801a52d939d4e6e3747720dd3b7/packages/%40aws-cdk/aws-codebuild/lib/project.ts#L944-L973), which is validated at deploy time).

eladb commented on packages/@aws-cdk/aws-codebuild/test/test.project.ts:
```diff
@@ -301,4 +301,34 @@ export = {
     test.done();
   },
+
+  'can use an ImmutableRole for a Project within a VPC'(test: Test) {
```
```suggestion
  'can use role.withoutPolicyUpdates() for a Project within a VPC'(test: Test) {
```

eladb commented on packages/@aws-cdk/aws-iam/lib/role.ts:
```diff
@@ -199,44 +216,11 @@ export class Role extends Resource implements IRole {
-      ? new MutableImport(scope, id)
-      : new ImmutableImport(scope, id);
+      ? new Import(scope, id)
+      : new ImmutableRole(new Import(scope, id));
```
should this be `new Import().withoutPolicyUpdates()`?

eladb commented on packages/@aws-cdk/aws-iam/lib/role.ts:
```diff
@@ -385,6 +369,19 @@ export class Role extends Resource implements IRole {
+   * If you do, you are responsible for adding the correct statements to the
+   * Role's policies yourself.
+   */
+  public withoutPolicyUpdates(): IRole {
```
Going back to our conversation, why can't this be in `IRole` and works both for `Role` and imported roles?

rix0rrr replied:
https://github.com/aws/aws-cdk/pull/5569#issuecomment-572538268

1. The interface is already starting to get onerous to implement and I don't want to make it worse. There are 2 different parts to it: "what makes an `IRole` a role", and "operations that can be done to any `IRole`, which are the same for every role anyway". This is part of the 2nd, but I'd argue those shouldn't actually be on `IRole` at all.

2. Even if we put it on `IRole`, my gut says the 99% case will be immutable, hence following up the import by calling this method. We want to optimize for the common case, and adding more verbiage seems the opposite. Either we (ALSO?) add a `withPolicyUpdates()` to switch it back on, or we just leave it as import properties for the `fromRoleArn()`. 

So either we make `IRole` *even heavier* or it doesn't need to be there at all.

eladb replied:
Sure. It’s not that critical.

charlocharlie commented:
It looks like this PR actually broke `{mutable: false}` for imported roles. Upgrading from 1.20.0 to 1.21.1 causes my imported roles to start taking on policies. I can create an issue if needed.

skinny85 commented:
> It looks like this PR actually broke `{mutable: false}` for imported roles. Upgrading from 1.20.0 to 1.21.1 causes my imported roles to start taking on policies. I can create an issue if needed.

Yes please!

charlocharlie commented:
Created issue: https://github.com/aws/aws-cdk/issues/5943
